lib/frontend/src/frontend.rs

```diff
@@ -614,8 +684,74 @@ impl<'a> FunctionBuilder<'a> {
             colocated: false,
         });
 
-        self.ins().call(libc_memmove, &[dest, source, num]);
+        self.ins().call(libc_memmove, &[dest, source, size]);
+    }
+
+    /// Optimised memmove for small moves.
+    pub fn emit_small_memmove(&mut self,
+                              isa: &TargetIsa,
+                              dest: Value,
+                              src: Value,
+                              size: u64,
+                              dest_align: u8,
+                              src_align: u8)
+    {
+        const THRESHOLD: u64 = 4;
+
+        assert!(src_align % 2 == 0,
+                "Alignment value for `src` is not a power of two.");
+
+        assert!(dest_align % 2 == 0,
+                "Alignment value for `dest` is not a power of two.");
+
+        macro_rules! fallback {
+            () => {{
+                let size_value = self.ins().iconst(isa.pointer_type(), size as i64);
+                self.call_memmove(isa, dest, src, size_value);
+            }}
+        }
+
+        if src_align == dest_align {
+            let min = ::std::cmp::min(src_align, dest_align) as u64;
+            let access_size = greatest_divisible_power_of_two(min, size);
+            let load_and_store_amount = size / access_size;
+
+            if load_and_store_amount > THRESHOLD {
+                fallback!();
+                return
+            }
+
+            let flags = MemFlags::new();
+            flags.aligned();
+
+            let registers: Vec<_> = (0..load_and_store_amount).map(|i| {
+                let offset = (access_size * i) as i32;
+                (self.ins().load(types::I8, flags, src, offset), offset)
+            }).collect();
+
+            for (value, offset) in registers {
+                self.ins().store(flags, value, dest, offset);
+            }
+        } else {
+            fallback!();
+        }
+    }
+
+}
+
+fn greatest_divisible_power_of_two(max: u64, multiple: u64) -> u64 {
+    let mut greatest_power = 2u64;
+    let mut next_power = greatest_power.next_power_of_two();
+
+    while max > next_power {
+        if (multiple / next_power) % 2 == 0 {
+            greatest_power = next_power;
+        }
+
+        next_power = next_power.next_power_of_two();
     }
```

Comment:
Power-of-two multiples are interesting because when you multiply a number by a power of 2, it's effectively just shifting the number left, inserting zeros at the low end. The number of zeros inserted is equal to the number of trailing zeros in the power of 2.

So instead of iterating here, we can use `trailing_zeros` to find the number of trailing zeros in the size. Its greatest power-of-two divisor is then the power of two with that many trailing zeros, or a one left-shifted by that number.